lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -170,6 +175,99 @@ def safe_remove_concurrent_index(table, options={})
+    if_not_exists: nil,
+    using: nil,
+    unique: nil,
+    where: nil,
+    comment: nil
+  )
+    raise ArgumentError, "Expected <name_suffix> to be present" unless name_suffix.present?
+
+    if ActiveRecord::Base.connection.postgresql_version < 11_00_00
+      raise PgHaMigrations::InvalidMigrationError, "Concurrent partitioned index creation not supported on Postgres databases before version 11"
+    end
+
+    parent_schema, parent_table = _schema_and_table_for(table)
+
+    fully_qualified_parent_table = "#{connection.quote_schema_name(parent_schema)}.#{parent_table}"
+    parent_index = "#{parent_table}_#{name_suffix}"
+
+    raise PgHaMigrations::InvalidMigrationError, "Table #{parent_table.inspect} is not a partitioned table" unless _partitioned_table?(parent_schema, parent_table)
+
+    _validate_index_name!(parent_index)
+
+    child_schemas_and_tables = connection.select_rows(<<~SQL)
+      SELECT child_ns.nspname, child.relname
+      FROM pg_inherits
+        JOIN pg_class parent        ON pg_inherits.inhparent = parent.oid
+        JOIN pg_class child         ON pg_inherits.inhrelid  = child.oid
+        JOIN pg_namespace parent_ns ON parent.relnamespace = parent_ns.oid
+        JOIN pg_namespace child_ns  ON child.relnamespace = child_ns.oid
+      WHERE parent.relname = #{connection.quote(parent_table)}
+        AND parent_ns.nspname = #{connection.quote(parent_schema)}
+    SQL
+
+    child_tables_with_metadata = child_schemas_and_tables.map do |child_schema, child_table|
+      raise PgHaMigrations::InvalidMigrationError, "Partitioned table #{parent_table.inspect} contains sub-partitions" if _partitioned_table?(child_schema, child_table)
+
+      child_index = "#{child_table}_#{name_suffix}"
+
+      _validate_index_name!(child_index)
+
+      [child_schema, child_table, child_index]
+    end
+
+    # TODO: take out ShareLock after issue #39 is implemented
+    safely_acquire_lock_for_table(fully_qualified_parent_table) do
+      # CREATE INDEX ON ONLY parent_table
+      unsafe_add_index(
+        fully_qualified_parent_table,
+        columns,
+        name: parent_index,
+        if_not_exists: if_not_exists,
+        using: using,
+        unique: unique,
+        where: where,
+        comment: comment,
+        algorithm: :only, # see lib/pg_ha_migrations/hacks/add_index_on_only.rb
+      )
+    end
+
+    child_tables_with_metadata.each do |child_schema, child_table, child_index|
+      # CREATE INDEX CONCURRENTLY ON child_table
+      safe_add_concurrent_index(
+        "#{connection.quote_schema_name(child_schema)}.#{child_table}",
+        columns,
+        name: child_index,
+        if_not_exists: if_not_exists,
+        using: using,
+        unique: unique,
+        where: where,
+      )
+    end
+
+    # Avoid taking out an unnecessary lock when there are
+    # no child tables or the index is already valid
+    if child_tables_with_metadata.present? && !_index_valid?(parent_schema, parent_index)
```

Comment:
This could happen if the initial method call falls for whatever reason and you re-execute with `if_not_exists: true`

Reply:
Do we have a test for that? Even if so, it seems like that would be better handled by checking to see if a valid index exists on the partition parent up-front rather than here at the end.

Reply:
That seems reasonable, but I know there was a reason why I did this... Lemme think about it and get back to you

Reply:
Yeah... I can't think of a reason why we wouldn't just do this check up-front. I thought that maybe there would be a situation where an index could exist with the same name (with different columns / uniqueness) and therefore we would want to let the `CREATE INDEX` calls fail. However, those calls will _not_ fail. It appears `IF NOT EXISTS` simply checks for the existence of an index matching the identifier

Reply:
Should be addressed in https://github.com/braintree/pg_ha_migrations/pull/85/commits/31cc86bd6194896a5cd88c77eea5b70586ed9132